Automattic/woocommerce-payments
# Continue optimization work for Frontend Currencies.

#### Changes proposed in this Pull Request

Further optimization of `FrontendCurrencies` based on the work of @millerf in https://github.com/Automattic/woocommerce-payments/pull/3450 . The purpose of the optimization was due to slow downs and timeouts within certain things in WordPress.com. I also created https://github.com/Automattic/woocommerce-payments/issues/3502 to further optimize the usage of `is_call_in_backtrace`.

* `Compatibility`
  *  Add a check in `init_compatibility_classes()` to make sure the classes are only initialized if there is more than 1 (the default) currency enabled.
  * Added `compatibility_classes` property and method to get the property to be able to do unit tests for the above mentioned check.
* `FrontendCurrencies`
  * Add new properties for caching purposes.
  * Add `selected_currency_changed()` method to reset cache properties.
  * Refactor existing methods to use or set cache properties.
  * Rename `should_override_currency_code` to `should_use_order_currency`.
  * Add `get_selected_currency_code` to get/set cache property value.
  * Refactor `should_use_order_currency` to check to make sure certain page is being viewed before calling intensive `is_call_in_backtrace``Utils` method.
* `MultiCurrency`
  * Reset cache in `FrontendCurrencies` when selected currency is changed.
* `Utils`
  * Add `is_page_with_vars` method that will check `$wp->query_vars` to see if a specific `pagename` is set and if any specific `vars` have been passed to it. This allows for checking for a specific page *before* calling `is_call_in_backtrace`.
* Add/update tests.

#### Testing instructions

* Unit tests should pass.
* Place an order for a product.
* On review page, change the selected currency.
* Order currency should remain the same.
* This also works by going to My Account > Orders, and then selecting a previous order.

-------------------
